### ISSUE TITLE
Linking errors on Scala 2.11 if an `object` is named `class`

### ISSUE THREAD

**Issue by Atry** [scala-js/scala-js#3888]
The following code works fine on Scala 2.12, but does not link on Scala 2.11:
``` scala
object `class` {
  def foo() = {
    println("Hello, World!")
  }
}

`class`.foo()
```

https://scalafiddle.io/sf/LRfrjg0/0

**Comment by Atry** [comment]
Note that this change breaks backward compatibility, so next version should be v2.0.0

**Comment by sjrd** [comment]
Some questions:

* Does it work if the object is in a package, not in an enclosing object? (ScalaFiddle always wraps its script in an object.)
* Does it work in the JVM? With/without wrapping in an enclosing object.
* Does it work in 1.x?

**Comment by Atry** [comment]
* JVM + Scala 2.11 - OK - https://scastie.scala-lang.org/CVfbzhAvTLSamyfW6G0gUA
* Scala.js 0.6 + Scala 2.12 - OK - https://scalafiddle.io/sf/OarwKFw/0
* Scala.js 0.6 + Scala 2.11 - error - https://scalafiddle.io/sf/LRfrjg0/0

Top level objects in a package should links but I did not test yet.

**Comment by sjrd** [comment]
OK I can also reproduce this in Scala.js 1.x + Scala 2.11. And it does work on the JVM, somehow. However, I'm very tempted to mark this as won't fix, because there *are* other problems with the same root cause on the JVM with nested classes/objects named `class`. For example:
```scala
scala> trait Foo { def foo: Int = 5 }; object Foo { class `class` { def bar: Int = 6 } }
<console>:11: warning: Class Foo$class differs only in case from Foo$class. Such classes will overwrite one another on case-insensitive filesystems.
       trait Foo { def foo: Int = 5 }; object Foo { class `class` { def bar: Int = 6 } }
             ^
defined trait Foo
defined object Foo

scala> new Foo {}.foo
java.lang.NoSuchMethodError: Foo$class.$init$(LFoo;)V
  ... 33 elided
```
The implementation class of `trait Foo` is called `Foo$class`, and the `class class` is also called `Foo$class`, so they clash at the JVM level and cause `LinkageError`s like that `NoSuchMethodError`.

I think we should just consider that a nested class/trait/object called `class` is invalid as implementation restriction for Scala 2.11 as a whole (JVM or JS) and that we don't want to go great lengths to kind-of support the bits that Scala/JVM tolerates.

**Comment by Atry** [comment]
It's OK to not fix it as Scala 2.11 is a legacy version. Just treat this issue as a backlog.

**Comment by gzm0** [comment]
Fixed in 36a92e3ff81789c57a6b82a866dcc0b1bc1b4653

**Comment by sjrd** [comment]
> Just treat this issue as a backlog.

We don't really do backlog for compiler bugs. ;) Either there is a very good reason not to fix something, and we close it as won't fix, or we fix it ASAP. And for the former case, there are exceptionally few issues. Overall there are [20 won't-fix bugs that are not internal nor 0.6.x-only](https://github.com/scala-js/scala-js/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed+label%3Abug+label%3Awontfix+-label%3Ainternal+-label%3A%220.6.x+only%22), and not all of those are compiler bugs. And with this issue fixed, we are back to 0 open compiler bugs.
